lib/OpenQA/Worker/Cache.pm

```diff
@@ -35,173 +37,325 @@ our (@ISA, @EXPORT);
     my $ua = Mojo::UserAgent->new(max_redirects => 2);
     $ua->max_response_size(0);
-    my $tx = $ua->build_tx(GET => sprintf '%s/tests/%d/asset/%s/%s', $host, $id, $type, basename($asset));
+    my $url = sprintf '%s/tests/%d/asset/%s/%s', $host, $id, $type, basename($asset);
+    print $log "Downloading " . basename($asset) . " from $url\n";
+    my $tx = $ua->build_tx(GET => $url);
+    my $headers;
 
     $ua->on(
         start => sub {
             my ($ua, $tx) = @_;
             my $progress     = 0;
             my $last_updated = time;
+            $tx->req->headers->header('If-None-Match' => qq{$etag}) if $etag;
             $tx->res->on(
                 progress => sub {
                     my $msg = shift;
+                    $msg->finish if $msg->code == 304;
                     return unless my $len = $msg->headers->content_length;
+
                     my $size = $msg->content->progress;
+                    $headers = $msg->headers if !$headers;
                     my $current = int($size / ($len / 100));
-                    local $| = 1;
                     # Don't spam the webui, update only every 5 seconds
                     if (time - $last_updated > 5) {
                         update_setup_status;
+                        toggle_asset_lock($asset, 1);
                         $last_updated = time;
                         if ($progress < $current) {
                             $progress = $current;
-                            print $log "Downloading $asset :", $size == $len ? 100 : $progress, "\n";
+                            print $log "CACHE: Downloading $asset :", $size == $len ? 100 : $progress . "\n";
                         }
                     }
                 });
         });
 
     $tx = $ua->start($tx);
-    if (!$tx->res->is_success) {
-        printf $log "CACHE: download of %s failed with %d: %s\n", basename($asset), $tx->res->code, $tx->res->message;
-        return undef;
+
+    if ($tx->res->code == 304) {
+        if (toggle_asset_lock($asset, 0)) {
+            print "CACHE: Content has not changed, not downloading the $asset but updating last use\n";
+        }
+        else {
+            print "CACHE: Abnormal situation, bailing out\n";
+            $asset = undef;
+        }
+    }
+    elsif ($tx->res->is_success) {
+        $etag = $headers->etag;
+        unlink($asset);
+        $asset = $tx->res->content->asset->move_to($asset)->path;
+        my $size = (stat $asset)[7];
+        if ($size == $headers->content_length) {
+            check_limits($size);
+            update_asset($asset, $etag, $size);
+            print $log "CACHE: Asset download sucessful to $asset, Cache size is: $cache_real_size\n";
+        }
+        else {
+            print $log "CACHE: Size of $asset differs, Expected: "
+              . $headers->content_length
+              . " / Downloaded: "
+              . $size . "\n";
+            $asset = undef;
```

Comment:
There's some handling for that on [lib/Worker/Engines/isotovideo.pm](https://github.com/os-autoinst/openQA/blob/master/lib/OpenQA/Worker/Engines/isotovideo.pm#L75) if asset is undef.